### DISCUSSION TITLE
boundary number error

### DISCUSSION THREAD

**Discussion by bmemlh** [lululxvi/deepxde#1788] [Q&A]
Hello everyone,

I have an error related to number of boundary points for the problem. If the boundary number is increased until approximately 100000, no error emerges. However, when the boundary number is slightly increased from this point an error message pops out, as shown below:

_

> ValueError                                Traceback (most recent call last)
> Cell In[13], line 10
>       8 bc_C_n2 = dde.icbc.NeumannBC(geomtime, lambda x:0.0, boundary5)
>       9 ic_C = dde.icbc.IC(geomtime, lambda x:C0/C_scale, initial, component=0)
> ---> 10 data = dde.data.TimePDE(geomtime, pde, [bc_C_d1, bc_C_d2, bc_C_d3, bc_C_n1, bc_C_n2, ic_C], num_domain=300000, num_boundary=110000, num_test=100000, num_initial=100000)
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:322, in TimePDE.__init__(self, geometryxtime, pde, ic_bcs, num_domain, num_boundary, num_initial, train_distribution, anchors, exclusions, solution, num_test, auxiliary_var_function)
>     306 def __init__(
>     307     self,
>     308     geometryxtime,
>    (...)
>     319     auxiliary_var_function=None,
>     320 ):
>     321     self.num_initial = num_initial
> --> 322     super().__init__(
>     323         geometryxtime,
>     324         pde,
>     325         ic_bcs,
>     326         num_domain,
>     327         num_boundary,
>     328         train_distribution=train_distribution,
>     329         anchors=anchors,
>     330         exclusions=exclusions,
>     331         solution=solution,
>     332         num_test=num_test,
>     333         auxiliary_var_function=auxiliary_var_function,
>     334     )
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:127, in PDE.__init__(self, geometry, pde, bcs, num_domain, num_boundary, train_distribution, anchors, exclusions, solution, num_test, auxiliary_var_function)
>     124 self.test_x, self.test_y = None, None
>     125 self.train_aux_vars, self.test_aux_vars = None, None
> --> 127 self.train_next_batch()
>     128 self.test()
> 
> File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:38, in run_if_all_none.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
>      36 x = [getattr(self, a) for a in attr]
>      37 if all(i is None for i in x):
> ---> 38     return func(self, *args, **kwargs)
>      39 return x if len(x) > 1 else x[0]
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:175, in PDE.train_next_batch(self, batch_size)
>     173 @run_if_all_none("train_x", "train_y", "train_aux_vars")
>     174 def train_next_batch(self, batch_size=None):
> --> 175     self.train_x_all = self.train_points()
>     176     self.bc_points()  # Generate self.num_bcs and self.train_x_bc
>     177     if self.bcs and config.hvd is not None:
> 
> File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:38, in run_if_all_none.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
>      36 x = [getattr(self, a) for a in attr]
>      37 if all(i is None for i in x):
> ---> 38     return func(self, *args, **kwargs)
>      39 return x if len(x) > 1 else x[0]
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:338, in TimePDE.train_points(self)
>     336 @run_if_all_none("train_x_all")
>     337 def train_points(self):
> --> 338     X = super().train_points()
>     339     if self.num_initial > 0:
>     340         if self.train_distribution == "uniform":
> 
> File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:38, in run_if_all_none.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
>      36 x = [getattr(self, a) for a in attr]
>      37 if all(i is None for i in x):
> ---> 38     return func(self, *args, **kwargs)
>      39 return x if len(x) > 1 else x[0]
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:265, in PDE.train_points(self)
>     263         tmp = self.geom.uniform_boundary_points(self.num_boundary)
>     264     else:
> --> 265         tmp = self.geom.random_boundary_points(
>     266             self.num_boundary, random=self.train_distribution
>     267         )
>     268     X = np.vstack((tmp, X))
>     269 if self.anchors is not None:
> 
> File ~\anaconda3\lib\site-packages\deepxde\geometry\timedomain.py:147, in GeometryXTime.random_boundary_points(self, n, random)
>     145 t = self.timedomain.random_points(n, random=random)
>     146 t = np.random.permutation(t)
> --> 147 return np.hstack((x, t))
> 
> File <__array_function__ internals>:180, in hstack(*args, **kwargs)
> 
> File ~\anaconda3\lib\site-packages\numpy\core\shape_base.py:345, in hstack(tup)
>     343     return _nx.concatenate(arrs, 0)
>     344 else:
> --> 345     return _nx.concatenate(arrs, 1)
> 
> File <__array_function__ internals>:180, in concatenate(*args, **kwargs)
> 
> ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 0, the array at index 0 has size 109999 and the array at index 1 has size 110000

_

In addition, when I assign the number of boundary points as 300000, the inconsistency between the matrices are varied as below:

> ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 0, the array at index 0 has size 299997 and the array at index 1 has size 300000

I don't have any idea why is this happening, please help me
Thank you!

**Comment by lululxvi** [comment]
It is strange. Could you show a minimal code example to reproduce the error?

**Comment by bmemlh** [comment]
Thank you for your interest Lu Lu!

Here is the relevant part of my code:

``
geom = dde.geometry.Rectangle(xmin=[0, 0], xmax=[10, 40])
timedomain = dde.geometry.TimeDomain(0, 2)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)
bc_C_d1 = dde.icbc.PointSetBC(coordinate_time_points, concentration_values, component=0)
bc_C_d2 = dde.icbc.DirichletBC(geomtime, lambda x:C0/C_scale, boundary1)
bc_C_d3 = dde.icbc.DirichletBC(geomtime, lambda x:C0/C_scale, boundary2)
bc_C_n1 = dde.icbc.NeumannBC(geomtime, lambda x:0.0, boundary3)
bc_C_n2 = dde.icbc.NeumannBC(geomtime, lambda x:0.0, boundary4)
ic_C = dde.icbc.IC(geomtime, lambda x:C0/C_scale, initial, component=0)
data = dde.data.TimePDE(geomtime, pde, [bc_C_d1, bc_C_d2, bc_C_d3, bc_C_n1, bc_C_n2, ic_C], num_domain=300000, num_boundary=120000, num_test=100000, num_initial=100000)
`
Solver runs perfectly when the num_boundary is lower than or equal to 100000 but when it increases to approximately 101000 the aforementioned error message emerges

Thanks again!

**Comment by lululxvi** [comment]
Could you send a **complete** but **simple** code so I can run and reproduce the error?

**Comment by bmemlh** [comment]
Thank you for your interest Lu Lu,

Here is the full code:

```
import deepxde as dde
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
import sys

R = 8.3145
Temp = 300.0
F = 96485.0
Ei = 0.5
Es = -0.5
Vs = 0.001
k0 = 0.050119
C0 = 1.0
L = 0.001
D = 1e-9
timemax = 2000.0
L_scale = L
D_scale = D
C_scale = C0
E_scale = (R*Temp)/F
Vs_scale = Vs
t_scale = (L*L)/D

def pde(x, C):
    dC_x = dde.grad.jacobian(C, x, j=0)
    dC_t = dde.grad.jacobian(C, x, j=2)
    dC_xx = dde.grad.hessian(C, x, i=0, j=0)
    dC_yy = dde.grad.hessian(C, x, i=1, j=1)
    NP = dC_t - (dC_xx+dC_yy)
    return NP

def boundary1(x, _):
    return np.isclose(x[0], 10)

def boundary2(x, _):
    return np.isclose(x[1], 40)

def boundary3(x, _):
    return np.greater_equal(x[0], 1) and np.less_equal(x[0], 10) and np.equal(x[1], 0)

def boundary4(x, _):
    return np.isclose(x[0], 0)

def initial(_, on_initial):
    return on_initial

time_points = np.array([])
a = np.array([0.0])
for i in range(1001):
    for j in range(201):
        time_points = np.append(time_points, a, axis=0)
    a = a + 0.002
time_points = time_points.reshape(time_points.shape[0], 1)

coordinate_points_x = np.array([])
for k in range(1001):
    coordinate_points_x = np.append(coordinate_points_x, np.arange(0, 1.005, 0.005), axis=0)
coordinate_points_y = np.repeat(0, 1001*201)
coordinate_points_x = coordinate_points_x.reshape(coordinate_points_x.shape[0], 1)
coordinate_points_y = coordinate_points_y.reshape(coordinate_points_y.shape[0], 1)
coordinate_time_points = np.concatenate([coordinate_points_x, coordinate_points_y, time_points], axis=1)

E1 = Ei/E_scale - ((Vs*t_scale)/E_scale)*time_points[0:100601]
E2 = Es/E_scale + ((Vs*t_scale)/E_scale)*(time_points[100601:201201]-(timemax/(2*t_scale)))
concentration1 = 1.0/(1.0+np.exp(-E_scale*(F/(R*Temp))*E1))
concentration2 = 1.0/(1.0+np.exp(-E_scale*(F/(R*Temp))*E2))
concentration_values = np.concatenate([concentration1, concentration2], axis=0)
concentration_values = concentration_values.reshape(concentration_values.shape[0], 1)

geom = dde.geometry.Rectangle(xmin=[0, 0], xmax=[10, 40])
timedomain = dde.geometry.TimeDomain(0, 2)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)
bc_C_d1 = dde.icbc.PointSetBC(coordinate_time_points, concentration_values, component=0)
bc_C_d2 = dde.icbc.DirichletBC(geomtime, lambda x:C0/C_scale, boundary1)
bc_C_d3 = dde.icbc.DirichletBC(geomtime, lambda x:C0/C_scale, boundary2)
bc_C_n1 = dde.icbc.NeumannBC(geomtime, lambda x:0.0, boundary3)
bc_C_n2 = dde.icbc.NeumannBC(geomtime, lambda x:0.0, boundary4)
ic_C = dde.icbc.IC(geomtime, lambda x:C0/C_scale, initial, component=0)
data = dde.data.TimePDE(geomtime, pde, [bc_C_d1, bc_C_d2, bc_C_d3, bc_C_n1, bc_C_n2, ic_C], num_domain=300000, num_boundary=150000, num_test=200000, num_initial=200000)

layer_size = [3]+[30]*3+[1]
activation = "tanh"
initializer = "Glorot normal"
net = dde.nn.FNN(layer_size, activation, initializer)
model = dde.Model(data, net)

resampler = dde.callbacks.PDEPointResampler(period=50)
model.compile("adam", lr=5e-4)
losshistory, train_state = model.train(iterations=20000, callbacks=[resampler])
dde.saveplot(losshistory, train_state, issave=False, isplot=True)
```

Thanks a lot again!
